src/rust/lib/graph-editor/src/component/node/port.rs

```diff
@@ -0,0 +1,417 @@
+
+        // This is a conditional rotation that allows the port to either point inwards
+        // or outwards.
+        let pi             = Var::from(PI);
+        let rotation_angle = is_inwards * pi;
+        let rotation_angle = Var::<Angle<Radians>>::from(rotation_angle);
+        let sculpted_shape = sculpted_shape.rotate(rotation_angle);
+
+        sculpted_shape.into()
+    }
+
+    ensogl::define_shape_system! {
+        (height:f32,width:f32,inner_radius:f32,is_inwards:f32) {
+            /// NOTE: `is_inwards` should only be 0.0 or 1.0.
+            new_port(height,width,inner_radius,is_inwards)
+        }
+    }
+
+    impl Shape{
+        /// Set the shape parameters derived from the `Specification`.
+        pub fn update_from_spec(&self,spec:&Specification){
+            self.height.set(spec.height);
+            self.inner_radius.set(spec.inner_radius);
+            self.width.set(spec.width.value.to_radians());
+            match &spec.direction{
+                Direction::In  => self.is_inwards.set(1.0),
+                Direction::Out => self.is_inwards.set(0.0),
+            };
+        }
+    }
+}
+
+
+
+// ============
+// === Port ===
+// ============
+
+/// Shape view for Input Port.
+#[derive(Debug,Clone,Copy)]
+pub struct InputPortView {}
+impl ShapeViewDefinition for InputPortView {
+    type Shape = shape::Shape;
+    fn new(shape:&Self::Shape, _scene:&Scene,_shape_registry:&ShapeRegistry) -> Self {
+        let spec  = Specification::default();
+        shape.update_from_spec(&spec);
+        shape.is_inwards.set(1.0);
+
+        let bbox = Vector2::new(2.0 * shape.height.get(), 2.0 * shape.height.get());
+        shape.sprite.size().set(bbox);
+
+        Self {}
+    }
+}
+
+/// Shape view for Input Port.
+#[derive(Debug,Clone,Copy)]
+pub struct OutputPortView {}
+impl ShapeViewDefinition for OutputPortView {
+    type Shape = shape::Shape;
+    fn new(shape:&Self::Shape, _scene:&Scene,_shape_registry:&ShapeRegistry) -> Self {
+        let spec  = Specification::default();
+        shape.update_from_spec(&spec);
+        shape.is_inwards.set(0.0);
+
+        let bbox = Vector2::new(2.0 * shape.height.get(), 2.0 * shape.height.get());
+        shape.sprite.size().set(bbox);
+
+        Self {}
+    }
+}
+
+/// Port definition.
+#[derive(Debug,Clone,CloneRef)]
+#[allow(missing_docs)]
+pub struct Port<T:ShapeViewDefinition+Clone> {
+    pub data : Rc<PortData<T>>
+}
+
+/// Type that represents an input port.
+pub type InputPort = Port<InputPortView>;
+
+/// Type that represents an output port.
+pub type OutputPort = Port<OutputPortView>;
+
+/// Internal data of `Port.
+#[derive(Debug,Clone)]
+#[allow(missing_docs)]
+pub struct PortData<T:ShapeViewDefinition+Clone> {
+        spec : RefCell<Specification>,
+    pub view : Rc<component::ShapeView<T>>
+}
+
+impl<T:ShapeViewDefinition<Shape=shape::Shape>+Clone> Port<T> {
+
+    /// Constructor.
+    pub fn new(spec:Specification) -> Self {
```

Comment:
When we are designing API one thing is super important - to keep constructors simple. This constructor should not have any parameters and it should just created port with default params. Then you should have an API to modify each parameter you want separately (see other comments here)

Reply:
I've added a default constructor and one that is `From<Specification>`. This function is now private.